controllers/profile_controller.go

```diff
@@ -79,43 +94,74 @@ func (r *ProfileSubscriptionReconciler) Reconcile(ctx context.Context, req ctrl.
 
 	pDef, err := git.GetProfileDefinition(pSub.Spec.ProfileURL, pSub.Spec.Branch, logger)
 	if err != nil {
-		r.patchStatusFailing(ctx, &pSub, logger, "error when fetching profile definition")
+		if err := r.patchStatus(ctx, &pSub, logger, readyFalse, "FetchProfileFailed", "error when fetching profile definition"); err != nil {
+			return ctrl.Result{}, err
+		}
 		return ctrl.Result{}, err
 	}
 
 	instance := profile.New(pDef, pSub, r.Client, logger)
-	err = instance.CreateArtifacts(ctx)
+	artifactStatus, err := instance.ArtifactStatus(ctx)
 	if err != nil {
-		r.patchStatusFailing(ctx, &pSub, logger, "error when creating profile artifacts")
 		return ctrl.Result{}, err
 	}
 
-	r.patchStatusRunning(ctx, &pSub, logger)
-	// TODO requeuing
-	return ctrl.Result{}, nil
-}
+	if !artifactStatus.ResourcesExist {
+		err = instance.CreateArtifacts(ctx)
+		if err != nil {
+			if err := r.patchStatus(ctx, &pSub, logger, readyFalse, "CreateFailed", "error when creating profile artifacts"); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, err
+		}
+		artifactStatus, err = instance.ArtifactStatus(ctx)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
 
-func (r *ProfileSubscriptionReconciler) patchStatusFailing(ctx context.Context, pSub *v1alpha1.ProfileSubscription, logger logr.Logger, message string) {
-	pSub.Status.State = "failing"
-	pSub.Status.Message = message
-	r.patchStatus(ctx, pSub, logger)
+	if len(artifactStatus.NotReadyConditions) == 0 {
+		if err := r.patchStatus(ctx, &pSub, logger, readyTrue, "ArtifactsReady", "all artifact resouces ready"); err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
+		var messages []string
+		status := readyUnknown
+		for _, condition := range artifactStatus.NotReadyConditions {
+			logger.Info(fmt.Sprintf("%s=%s, message:%s", condition.Type, string(condition.Status), condition.Message))
+			messages = append(messages, condition.Message)
+			if string(condition.Status) == readyFalse {
+				status = readyFalse
+			}
+		}
+		if err := r.patchStatus(ctx, &pSub, logger, status, "ArtifactNotReady", strings.Join(messages, ",")); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
```

Comment:
WDYT about dropping the else and replacing it with a return in the len check if? That way the flow of the function is not broken.

Reply:
sounds good!